### ISSUE TITLE
Constructors with keyword arguments a lot slower than ones without them

### ISSUE THREAD

**Issue by wsshin** [JuliaLang/julia#19668]
I find constructors taking keyword arguments are a lot slower than constructors that don't.  For example,
```
julia> immutable PositionalArg
           x::Int
           PositionalArg(x::Int=1) = new(x)
       end

julia> immutable KeywordArg
           x::Int
           KeywordArg(;x::Int=1) = new(x)
       end

julia> using BenchmarkTools

julia> @benchmark PositionalArg()
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     2.087 ns (0.00% GC)
  median time:      2.092 ns (0.00% GC)
  mean time:        2.381 ns (0.00% GC)
  maximum time:     386.548 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark KeywordArg()
BenchmarkTools.Trial:
  memory estimate:  16.00 bytes
  allocs estimate:  1
  --------------
  minimum time:     75.253 ns (0.00% GC)
  median time:      78.912 ns (0.00% GC)
  mean time:        101.540 ns (1.62% GC)
  maximum time:     26.617 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     970
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
Such a huge difference seems to be caused by the additional allocations that constructors with keyword arguments use.  However, I'm not sure if that has to be the case, because ordinary functions don't seem to be using additional allocations when taking keyword arguments:
```
julia> posarg(x::Int=1) = x
posarg (generic function with 2 methods)

julia> kwdarg(;x::Int=1) = x
kwdarg (generic function with 1 method)

julia> @benchmark posarg()
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.997 ns (0.00% GC)
  median time:      1.998 ns (0.00% GC)
  mean time:        2.282 ns (0.00% GC)
  maximum time:     95.239 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark kwdarg()
BenchmarkTools.Trial:
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.997 ns (0.00% GC)
  median time:      1.998 ns (0.00% GC)
  mean time:        2.042 ns (0.00% GC)
  maximum time:     273.882 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1000
  time tolerance:   5.00%
  memory tolerance: 1.00%
```
Is this a matter that can be fixed easily, or I need to refrain from using keyword arguments in constructors?

Here is the version info:
```
julia> versioninfo()
Julia Version 0.6.0-dev.1636
Commit 99b6a8c* (2016-12-20 13:46 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.0.0)
  CPU: Intel(R) Core(TM)2 Duo CPU     T9900  @ 3.06GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Penryn)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, penryn)
```

**Comment by KristofferC** [comment]
Duplicate of https://github.com/JuliaLang/julia/issues/9551?

**Comment by wsshin** [comment]
Seems to be related to that issue, but then shouldn't ordinary functions suffer from the same problem?   My example above shows that ordinary functions with keyword arguments do not suffer from the problem.

In that issue, https://github.com/JuliaLang/julia/issues/9551#issuecomment-73364527 suspects that what is slow on keyword arguments is not the function itself, but the dispatch system.  If that's the case, then both constructors and ordinary functions should suffer from performance degradation when keyword arguments are used, but my example shows that is not the case...

**Comment by KristofferC** [comment]
Looks like julia inlines the keyword in case it is not being used:
```
julia> @code_llvm kwdarg()

define i64 @julia_kwdarg_71585() #0 {
top:
  ret i64 1
}
```

```
julia> @code_llvm kwdarg(x=2)
define i64 @"julia_#kwdarg_71593"(%jl_value_t*) #0 {
top:
  %ptls_i8 = call i8* asm "movq %fs:0, $0;\0Aaddq $$-2672, $0", "=r,~{dirflag},~{fpsr},~{flags}"() #2
  %ptls = bitcast i8* %ptls_i8 to %jl_value_t***
  %1 = alloca [13 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [13 x %jl_value_t*], [13 x %jl_value_t*]* %1, i64 0, i64 0
  %2 = getelementptr [13 x %jl_value_t*], [13 x %jl_value_t*]* %1, i64 0, i64 11
  %3 = getelementptr [13 x %jl_value_t*], [13 x %jl_value_t*]* %1, i64 0, i64 2
  %4 = getelementptr [13 x %jl_value_t*], [13 x %jl_value_t*]* %1, i64 0, i64 10
  %5 = bitcast %jl_value_t** %2 to i8*
etc etc 
```

**Comment by wsshin** [comment]
Are you saying that inlining occurs for ordinary functions, but not for constructors, when functions/constructors are defined with keyword arguments but those keyword arguments are not used?

Some sort of code simplification seems to be occurring for constructors as well, because `@code_llvm KeywordArg()` spits out much shorter code than `@code_llvm KeywordArg(x=1)`.  

Still, `@code_llvm PositionalArg()` is much shorter than `@code_llvm KeywordArg(x=1)`:
```
julia> @code_llvm PositionalArg()

define %PositionalArg @julia_Type_60674(%jl_value_t*) #0 {
top:
  ret %PositionalArg { i64 1 }

julia> @code_llvm KeywordArg()

define %jl_value_t* @julia_Type_60675(%jl_value_t*, %jl_value_t**, i32) #0 {
top:
  %3 = alloca %jl_value_t**, align 8
  store volatile %jl_value_t** %1, %jl_value_t*** %3, align 8
  %4 = call %jl_value_t*** @jl_get_ptls_states() #4
  %5 = alloca [6 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [6 x %jl_value_t*], [6 x %jl_value_t*]* %5, i64 0, i64 0
  %6 = getelementptr [6 x %jl_value_t*], [6 x %jl_value_t*]* %5, i64 0, i64 3
  %7 = getelementptr [6 x %jl_value_t*], [6 x %jl_value_t*]* %5, i64 0, i64 2
  %8 = bitcast %jl_value_t** %6 to i8*
  call void @llvm.memset.p0i8.i32(i8* %8, i8 0, i32 24, i32 8, i1 false)
  %9 = bitcast [6 x %jl_value_t*]* %5 to i64*
  store i64 8, i64* %9, align 8
  %10 = bitcast %jl_value_t*** %4 to i64*
  %11 = load i64, i64* %10, align 8
  %12 = getelementptr [6 x %jl_value_t*], [6 x %jl_value_t*]* %5, i64 0, i64 1
  %13 = bitcast %jl_value_t** %12 to i64*
  store i64 %11, i64* %13, align 8
  store %jl_value_t** %.sub, %jl_value_t*** %4, align 8
  store %jl_value_t* null, %jl_value_t** %7, align 8
  %14 = load %jl_value_t*, %jl_value_t** inttoptr (i64 4677303264 to %jl_value_t**), align 32
  %15 = icmp eq %jl_value_t* %14, null
  br i1 %15, label %fail, label %pass

fail:                                             ; preds = %top
  call void @jl_throw(%jl_value_t* inttoptr (i64 4595213136 to %jl_value_t*))
  unreachable

pass:                                             ; preds = %top
  %16 = getelementptr [6 x %jl_value_t*], [6 x %jl_value_t*]* %5, i64 0, i64 4
  %17 = getelementptr [6 x %jl_value_t*], [6 x %jl_value_t*]* %5, i64 0, i64 5
  store %jl_value_t* %14, %jl_value_t** %7, align 8
  store %jl_value_t* %14, %jl_value_t** %6, align 8
  store %jl_value_t* inttoptr (i64 4593893536 to %jl_value_t*), %jl_value_t** %16, align 8
  store %jl_value_t* inttoptr (i64 4677610672 to %jl_value_t*), %jl_value_t** %17, align 8
  %18 = call %jl_value_t* @jl_apply_generic(%jl_value_t** %6, i32 3)
  %19 = load i64, i64* %13, align 8
  store i64 %19, i64* %10, align 8
  ret %jl_value_t* %18
}
```

 It seems like inlining is applied in `PositionalArg()`, `posarg()`, `kwdarg()`, but not in `KeywordArg()`.  Why?

**Comment by yuyichao** [comment]
Again please don't use code_llvm when you can't understand it and the problem can be shown at a much higher level!!

```
julia> immutable KWArg
           x::Int
           KWArg(;x::Int=1) = new(x)
       end

julia> kwarg(;x::Int=1) = x
kwarg (generic function with 1 method)

julia> @code_lowered KWArg()
CodeInfo(:(begin 
        nothing
        return ((Core.getfield)($(QuoteNode(Core.Box(#call#3))),:contents))(1,#self#)
    end))

julia> @code_lowered kwarg()
CodeInfo(:(begin 
        nothing
        return (Main.#kwarg#5)(1,#self#)
    end))
```

**Comment by KristofferC** [comment]
I used it as a "it inlines here" vs "it gets messy" here!! I did not try do use it for any further analysis!!

**Comment by yuyichao** [comment]
> I used it as a "it inlines here" vs "it gets messy" here!! 

Use `@code_warntype`

```jl
julia> @code_warntype kwarg()
Variables:
  #self#::#kwarg

Body:
  begin 
      return 1
  end::Int64

julia> @code_warntype KWArg()
Variables:
  #self#::Type{KWArg}

Body:
  begin 
      return ((Core.getfield)($(QuoteNode(Core.Box(#call#3))),:contents)::Any)(1,#self#::Type{KWArg})::Any
  end::Any
```

**Comment by yuyichao** [comment]
Also, the keyword is actually inlined, as shown in the lowered code. What's not inlined is the call to the inner function. And

> I did not try do use it for any further analysis!!

Sure, but it had clearly mislead @wsshin to use the wrong tool for analysis.

**Comment by wsshin** [comment]
@KristofferC @yuyichao Thanks for pointing out my negligent use of `@code_llvm`!  Both of you seem aware that `@code_llvm` is not a correct tool for analysis here, but I wasn't.  I tried `@code_warntype` before `@code_llvm`, but to be honest I failed to spot from the `@code_warntype` result that the keyword arguments were inlined for the constructor call as well.  I only realized this after reading @yuyichao's last comment.

So, is the source of the problem that @yuyichao mentioned (that the call to the inner function is not inlined inside the constructor) something that can be fixed later?  I'm curious.
